backend/chainlit/data/sql_alchemy.py

```diff
@@ -0,0 +1,496 @@
+                    start = i + 1
+                    break
+        end = start + pagination.first
+        paginated_threads = filtered_threads[start:end] or []
+
+        has_next_page = len(filtered_threads) > end
+        start_cursor = paginated_threads[0]['id'] if paginated_threads else None
+        end_cursor = paginated_threads[-1]['id'] if paginated_threads else None
+
+        return PaginatedResponse(
+            pageInfo=PageInfo(hasNextPage=has_next_page, startCursor=start_cursor, endCursor=end_cursor),
+            data=paginated_threads
+        )
+    
+    ###### Steps ######
+    @queue_until_user_message()
+    async def create_step(self, step_dict: 'StepDict'):
+        logger.info(f"SQLAlchemy: create_step, step_id={step_dict.get('id')}")
+        if not getattr(context.session.user, 'id', None):
+            raise ValueError("No authenticated user in context")
+        step_dict['showInput'] = str(step_dict.get('showInput', '')).lower() if 'showInput' in step_dict else None
+        parameters = {key: value for key, value in step_dict.items() if value is not None and not (isinstance(value, dict) and not value)}
+        parameters['metadata'] = json.dumps(step_dict.get('metadata', {}))
+        columns = ', '.join(f'"{key}"' for key in parameters.keys())
+        values = ', '.join(f':{key}' for key in parameters.keys())
+        updates = ', '.join(f'"{key}" = :{key}' for key in parameters.keys() if key != 'id')
+        query = f"""
+            INSERT INTO steps ({columns})
+            VALUES ({values})
+            ON CONFLICT (id) DO UPDATE
+            SET {updates};
+        """
+        await self.execute_sql(query=query, parameters=parameters)
+    
+    @queue_until_user_message()
+    async def update_step(self, step_dict: 'StepDict'):
+        logger.info(f"SQLAlchemy: update_step, step_id={step_dict.get('id')}")
+        await self.create_step(step_dict)
+
+    @queue_until_user_message()
+    async def delete_step(self, step_id: str):
+        logger.info(f"SQLAlchemy: delete_step, step_id={step_id}")
+        # Delete feedbacks/elements/steps
+        feedbacks_query = """DELETE FROM feedbacks WHERE "forId" = :id"""
+        elements_query = """DELETE FROM elements WHERE "forId" = :id"""
+        steps_query = """DELETE FROM steps WHERE "forId" = :id"""
+        parameters = {"id": step_id}
+        await self.execute_sql(query=feedbacks_query, parameters=parameters)
+        await self.execute_sql(query=elements_query, parameters=parameters)
+        await self.execute_sql(query=steps_query, parameters=parameters)
+    
+    ###### Feedback ######
+    async def upsert_feedback(self, feedback: Feedback) -> str:
+        logger.info(f"SQLAlchemy: upsert_feedback, feedback_id={feedback.id}")
+        if not getattr(context.session.user, 'id', None):
+            raise ValueError("No authenticated user in context")
```

Comment:
You need to remove this check. The `upsert_feedback` is only called from a route, so it's outside of the user session (so attempting to use `context` will always through).

I don't have a good solution to prevent adding feedback to the DB for non-authenticated chainlit apps for now.

```suggestion
```

Reply:
Maybe we're overthinking this. I don't see a scenario where authenticated and non-authenticated users coexist in the same app/db. With authentication enabled users can't use my app with authentication. If someone doesn't have authentication set up I don't know why they'd set up a custom data layer.

I just don't see how this scenario we're trying to guard against would be possible, but maybe I'm missing something?

Reply:
Yes this is a edge edge case I agree. I can sadly think about a least one scenario where it could happen: a app owner disables the authentication for a few weeks.
To be fair this is the only thing that I've found in today's testing session, and getting this fixed would enable me to do a last code review / testing session.

An option that I've just though about is to use the `require_login` method instead of checking the context (https://github.com/Chainlit/chainlit/blob/59398777b7f04048ddd8961a47924a6a58c23be0/backend/chainlit/auth.py#L31-L37).

Reply:
I like this idea! Then we can just block initialzation of the custom data layer itself if no authentication. Great solution!

Where would we call require_login()? I've tried the app.py before the decorators and in sqlalchemy.py but it is returning `None` for me even when I do have authentication enabled.

```python
from chainlit.auth import require_login
is_authentication_enabled = require_login()

class SQLAlchemyDataLayer(BaseDataLayer):
    def __init__(self, conninfo: str, ssl_require: bool = False, storage_provider: Optional[BaseStorageClient] = None, user_thread_limit: Optional[int] = 1000):
        if not is_authentication_enabled:
            print(f'is_authentication_enabled: {is_authentication_enabled}')
            raise PermissionError("Authentication is required to use SQLAlchemyDataLayer")
```